components/server/src/auth/verification-service.ts

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+import { User, AdditionalUserData } from "@gitpod/gitpod-protocol";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { inject, injectable, postConstruct } from "inversify";
+import { Config } from "../config";
+import { Twilio } from "twilio";
+import { ServiceContext } from "twilio/lib/rest/verify/v2/service";
+import { WorkspaceDB } from "@gitpod/gitpod-db/lib";
+import { ConfigCatClientFactory } from "@gitpod/gitpod-protocol/lib/experiments/configcat-server";
+
+@injectable()
+export class VerificationService {
+    @inject(Config) protected config: Config;
+    @inject(WorkspaceDB) protected workspaceDB: WorkspaceDB;
+    @inject(ConfigCatClientFactory) protected readonly configCatClientFactory: ConfigCatClientFactory;
+
+    protected verifyService: ServiceContext;
+
+    @postConstruct()
+    protected initialize(): void {
+        if (this.config.twilioConfig) {
+            const client = new Twilio(this.config.twilioConfig.accountSID, this.config.twilioConfig.authToken);
+            this.verifyService = client.verify.v2.services(this.config.twilioConfig.serviceID);
+        }
+    }
+
+    public async needsVerification(user: User): Promise<boolean> {
+        if (!this.config.twilioConfig) {
+            return false;
+        }
+        if (!!user.additionalData?.lastVerificationTime) {
+            return false;
+        }
+        // we treat existing users (created before we introduced phone vwerification) as verified
+        if (user.creationDate < "2022-08-22") {
+            return false;
+        }
+        const isPhoneVerificationEnabled = await this.configCatClientFactory().getValueAsync(
+            "isPhoneVerificationEnabled",
+            false,
+            {
+                user,
+            },
+        );
+        return isPhoneVerificationEnabled;
+    }
+
+    public markVerified(user: User): User {
+        AdditionalUserData.set(user, {
+            lastVerificationTime: new Date().toISOString(),
+        });
+        return user;
+    }
+
+    public async sendVerificationToken(phoneNumber: string): Promise<void> {
+        if (!this.verifyService) {
+            throw new Error("No verification service configured.");
+        }
+        const verification = await this.verifyService.verifications.create({ to: phoneNumber, channel: "sms" });
+        log.info("Verification code sent", { phoneNumber, status: verification.status });
+    }
+
+    public async verifyVerificationToken(phoneNumber: string, oneTimePassword: string): Promise<boolean> {
+        if (!this.verifyService) {
+            throw new Error("No verification service configured.");
+        }
+        const verification_check = await this.verifyService.verificationChecks.create({
```

Comment:
try-catch, otherwise it's handled as internal server error whenever one uses the happy finger on the send button.

Reply:
I don't get this comment. Why is there always an internal server error?

Reply:
as this is not wrapped in a `ResponseError` it'll be logged automatically as an internal server error, thus it's loud on the logs, and it seems not to be rendered properly on frontend, at least I couldn't see twillios rate limiting error rendered on frontend.

Reply:
It is fine to be loud on the logs. It's not a user error when an error is thrown but something with the system. The error is thrown to frontend and I'll make it so that the user gets feedback about the error.